# What's in the Smart-seq2 Multi Sample Pipeline Loom File?

The Loom file is an HDF5 file generated using [Loompy v.3.0.6](http://loompy.org/). It contains global attributes detailing how counts were generated for the single-cell or single-nuclei parameters ([Table 1](#table-1-global-attributes)). It additionally contains UMI-corrected counts as well as multiple metrics for both individual cells (the columns of the matrix; [Table 2](#table-2-column-attributes-cell-metrics)) and individual genes (the rows of the matrix; [Table 3](#table-3-row-attributes-gene-metrics)). The tables below document these metrics, list which tools generate them, and define them. This Loom file is the default matrix output of the Optimus pipeline.  

**Note**: Loom files generated by the Smart-seq2 Multi Sample are different from the final Loom file distributed on the [Human Cell Atlas Data Portal](https://data.humancellatlas.org/explore/projects), which removes some of the metadata detailed in this document and contains additional metadata relating to each individual project. 

## Table 1. Global Attributes
The global attributes in the Loom apply to the whole file, not any specific part. There are two global attributes for the Optimus Loom. 

| Attribute | Details |
| :-- | :-- |
| LOOM_SPEC_VERSION | String with the loom file spec version |
| CreationDate | String describing if the pipeline counts exonic or whole transcript (exonic and intronic) reads. For the single-cell mode (counting_mode = sc_rna), the value will be "exonic"; for the single-nuclei mode (counting_mode = sn_rna), the value will be "whole_transcript" |
| sample_id | The sample or cell id listed in the pipeline configuration file. This can be any string, but we recommend it be consistent with any sample metadata. |
 

## Table 2. Column Attributes (Cell Metrics) 

| Cell Metrics | Program |Details             |
|:---|:---:|:--------------------|
| `ACCUMULATION_LEVEL` |
| `ALIGNED_READS` |
| `AT_DROPOUT` |
| `Aligned 0 time` |
| `Aligned 1 time` |
| `Aligned >1 times` |
| `BAD_CYCLES.UNPAIRED` |
| `CODING_BASES` |
| `CORRECT_STRAND_READS` |
| `ESTIMATED_LIBRARY_SIZE` |
| `GC_DROPOUT` |
| `GC_NC_0_19` |
| `GC_NC_20_39` |
| `GC_NC_40_59` |
| `GC_NC_60_79` |
| `GC_NC_80_100` |
| `IGNORED_READS` |
| `INCORRECT_STRAND_READS` |
| `INTERGENIC_BASES` |
| `INTRONIC_BASES` |
| `MEAN_READ_LENGTH.UNPAIRED` |
| `MEDIAN_3PRIME_BIAS` |
| `MEDIAN_5PRIME_BIAS` |
| `MEDIAN_5PRIME_TO_3PRIME_BIAS` |
| `MEDIAN_CV_COVERAGE` |
| `NUM_R1_TRANSCRIPT_STRAND_READS` |
| `NUM_R2_TRANSCRIPT_STRAND_READS` |
| `NUM_UNEXPLAINED_READS` |
| `Overall alignment rate` |
| `PCT_ADAPTER.UNPAIRED` |
| `PCT_CHIMERAS.UNPAIRED` |
| `PCT_CODING_BASES` |
| `PCT_CORRECT_STRAND_READS` |
| `PCT_INTERGENIC_BASES` |
| `PCT_INTRONIC_BASES` |
| `PCT_MRNA_BASES` |
| `PCT_PF_READS.UNPAIRED` |
| `PCT_PF_READS_ALIGNED.UNPAIRED` |
| `PCT_PF_READS_IMPROPER_PAIRS.UNPAIRED` |
| `PCT_R1_TRANSCRIPT_STRAND_READS` |
| `PCT_R2_TRANSCRIPT_STRAND_READS` |
| `PCT_READS_ALIGNED_IN_PAIRS.UNPAIRED` |
| `PCT_RIBOSOMAL_BASES` |
| `PCT_USABLE_BASES` |
| `PCT_UTR_BASES` |
| `PERCENT_DUPLICATION` |
| `PF_ALIGNED_BASES` |
| `PF_ALIGNED_BASES.UNPAIRED` |
| `PF_BASES` |
| `PF_HQ_ALIGNED_BASES.UNPAIRED` |
| `PF_HQ_ALIGNED_Q20_BASES.UNPAIRED` |
| `PF_HQ_ALIGNED_READS.UNPAIRED` |
| `PF_HQ_ERROR_RATE.UNPAIRED` |
| `PF_HQ_MEDIAN_MISMATCHES.UNPAIRED` |
| `PF_INDEL_RATE.UNPAIRED` |
| `PF_MISMATCH_RATE.UNPAIRED` |
| `PF_NOISE_READS.UNPAIRED` |
| `PF_READS.UNPAIRED` |
| `PF_READS_ALIGNED.UNPAIRED` |
| `PF_READS_IMPROPER_PAIRS.UNPAIRED` |
| `READS_ALIGNED_IN_PAIRS.UNPAIRED` |
| `READS_USED` |
| `READ_PAIRS_EXAMINED` |
| `READ_PAIR_DUPLICATES` |
| `READ_PAIR_OPTICAL_DUPLICATES` |
| `RIBOSOMAL_BASES` |
| `SECONDARY_OR_SUPPLEMENTARY_RDS` |
| `STRAND_BALANCE.UNPAIRED` |
| `TOTAL_CLUSTERS` |
| `TOTAL_READS.UNPAIRED` |
| `Total reads` |
| `UNMAPPED_READS` |
| `UNPAIRED_READS_EXAMINED` |
| `UNPAIRED_READ_DUPLICATES` |
| `UTR_BASES` |
| `WINDOW_SIZE` |
| `alignable reads` |
| `cell_names` |
| `filtered reads` |
| `multiple mapped` |
| `strand` |
| `total alignments` |
| `total reads` |
| `unalignable reads` |
| `uncertain reads` |
| `unique aligned` |


## Table 3. Row Attributes (Gene Metrics)

| Gene Metrics                  | Program            |Details                 | 
|-------------------------------|--------------------|------------------------|
|`ensembl_ids` | [GENCODE GTF](https://www.gencodegenes.org/) | The gene_id listed in the GENCODE GTF |
|`gene_names` | [GENCODE GTF](https://www.gencodegenes.org/) | The unique gene_name provided in the GENCODE GTF |

