| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |
| :----: | :---: | :----: | :--------------: |
| v.1.0.0 | December, 2019 | [Elizabeth Kiernan](mailto:ekiernan@broadinstitute.org) | Please file GitHub issues in skylab or contact [Kylee Degatano](mailto:kdegatano@broadinstitute.org) |

# Smart-seq2 Multi Sample Pipeline
## Introduction
The Smart-seq2 Multi Sample pipeline is a wrapper around the [Smart-seq2 Single Sample](https://github.com/HumanCellAtlas/skylab/tree/master/pipelines/smartseq2_single_sample) pipeline. It is developed by the Data Coordination Platform of the Human Cell Atlas to process single-cell RNAseq (scRNAseq) data generated by Smart-seq2 assays. The workflow processes multiple cells by importing and running the [Smart-seq2 Single Sample workflow](https://github.com/HumanCellAtlas/skylab/blob/master/pipelines/smartseq2_single_sample/SmartSeq2SingleSample.wdl) for each individual cell and then merging the resulting Zarr output files and optionally generating a Loom file. 

Full details about the Smart-seq2 pipeline can be read in the [Smart-seq2 Single Sample workflow documentation](https://github.com/HumanCellAtlas/skylab/tree/master/pipelines/smartseq2_single_sample) in GitHub. 
 
## Inputs 

The [Smartseq2PlateExample.json](SmartSeq2PlateExample.json) file provides example inputs for the Smart-seq2 Multi Sample workflow. 

### Sample Data Input
The sample data inputs are similar to those specified in the [Smart-seq2 Single Sample README](https://github.com/HumanCellAtlas/skylab/blob/master/pipelines/smartseq2_single_sample/README.md). The Smart-seq2 Multi Sample workflow processes both single- and paired-end samples; however, these samples can not be mixed in the same run. 
For the Multi Sample workflow, fastq files must be located in a cloud-accesible folder and named with the following convention:

*  sample_filename_prefix\_1.fastq (forward reads for paired-end samples or reads for single-end samples)
*  sample_filename_prefix\_2.fastq (reverse reads for paired-end samples and not applicable to single-end samples)

The Multi Sample workflow does not require a list of all fastq files; instead, it accepts a list (array of strings) of each sample's filename prefix. Additionally, the workflow accepts the URL for the fastq folder location as a string (see table below for details). Then, the workflow uses the filename prefixes and the URL string to find the appropriate input files and run subsequent analyses.


### Additional Input
The reference inputs are identical to those specified in the "Additional Reference Inputs" section of the [Smart-seq2 Single Sample README](https://github.com/HumanCellAtlas/skylab/blob/master/pipelines/smartseq2_single_sample/README.md). 

In addition to the Smart-seq2 Single Sample inputs, the Multi Sample workflow requires the inputs listed in the following table:

| Input name | Input Description | Input Type |
| --- | --- | --- |
| file_prefix | The URL path to the fastq file folder | String | 
| input_file_names | Array of filename prefixes; workflow will append with \_1.fastq.gz and \_2.fastq.gz | Array of Strings |
| bacth_id | Identifier for the batch of multiple samples | String |
| output_loom | Boolean description for optional Loom output | Boolean |

 
 ## Smart-seq2 Multi Sample Task Summary
 The Smart-seq2 Multi Sample pipeline calls three tasks:
 
 1) [SmartSeq2SingleSample.wdl](https://github.com/HumanCellAtlas/skylab/blob/master/pipelines/smartseq2_single_sample/SmartSeq2SingleSample.wdl): a task that runs the Smart-seq2 Single Sample workflow
 2) [SmartSeq2PlateAggregation.wdl](https://github.com/HumanCellAtlas/skylab/blob/master/library/tasks/SmartSeq2PlateAggregation.wdl): the wrapper pipeline that checks the inputs, initiates per cell processing and aggregates the results
 3) [ZarrUtils.wdl](https://github.com/HumanCellAtlas/skylab/blob/master/library/tasks/ZarrUtils.wdl): an optional task that converts the Zarr array to Loom file format 
 
 ## Outputs
 
 The following table lists the outputs of the Smart-seq2 Multi Sample workflow.
 
 | Output file name | Output Description | Output Type |
 | --- | --- | --- |
 | bam_files | An array of genome-aligned BAM files (one for each sample) generated with HISAT2  | Array |
 | bam_index_files |  An array of BAM index files generated with HISAT2 | Array |
 | zarrout | Array of Zarr output files for all samples  | Array |
 | loom_output | An optional Loom file  | File |
 
The Zarr outputs and BAM files are described in the [Smart-seq2 Single Sample README](https://github.com/HumanCellAtlas/skylab/blob/master/pipelines/smartseq2_single_sample/README.md).
 
 # Versioning

Release information for the Smart-seq2 Multi Sample Pipeline can be found in the [changelog](https://github.com/HumanCellAtlas/skylab/blob/master/pipelines/smartseq2_multisample/MultiSampleSmartSeq2.changelog.md). Please note that any major changes to the Smart-seq2 pipeline will be documented in the [Smart-seq2 Single Sample changelog](https://github.com/HumanCellAtlas/skylab/blob/master/pipelines/smartseq2_single_sample/SmartSeq2SingleSample.changelog.md).  
 
 
 



