| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |
| :----: | :---: | :----: | :--------------: |
| v.1.0.0 | December, 2019 | [Elizabeth Kiernan](mailto:ekiernan@broadinstitute.org) | Please file GitHub issues in skylab or contact [Kylee Degatano](mailto:kdegatano@broadinstitute.org) |

# Smart-seq2 Multi Sample Pipeline
## Introduction
The Smart-seq2 Multisample pipeline is modified from the Smart-seq2 Single Sample pipeline. It is designed by the Data Coordination Platform of the Human Cell Atlas to process single-cell RNAseq (scRNAseq) data generated by Smart-seq2 assays. The workflow processes multiple cells by importing the Smart-seq2 Single cell workflow for each indvidual cell and then merging the resulting Zarr output files. 

More details about the Smart-seq2 pipeline can be read in the [Smart-seq2 Single Sample worflow documentation](https://github.com/HumanCellAtlas/skylab/tree/master/pipelines/smartseq2_single_sample) in GitHub. 
 
# Inputs
 - Inputs are identical for the SS2 pipeline however, for running 
 multiple samples a URL prefix is provided along with an array of filename prefixes 
 under that path. File names are autogenerated by appending _1.fastq.gz and (in the case of 
 paired end) _2.fastq.gz at the end of the filenames
 - Note that single and paired end samples can not be mixed in the same run
 
 # Versioning
 
