| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |
| :----: | :---: | :----: | :--------------: |
| v.1.0.0 | December, 2019 | [Elizabeth Kiernan](mailto:ekiernan@broadinstitute.org) | Please file GitHub issues in skylab or contact [Kylee Degatano](mailto:kdegatano@broadinstitute.org) |

# Smart-seq2 Multi Sample Pipeline
## Introduction
The Smart-seq2 Multi Sample pipeline is modified from the [Smart-seq2 Single Sample](https://github.com/HumanCellAtlas/skylab/tree/master/pipelines/smartseq2_single_sample) pipeline. It is designed by the Data Coordination Platform of the Human Cell Atlas to process single-cell RNAseq (scRNAseq) data generated by Smart-seq2 assays. The workflow processes multiple cells by importing and running the [Smart-seq2 Single cell workflow](https://github.com/HumanCellAtlas/skylab/blob/master/pipelines/smartseq2_single_sample/SmartSeq2SingleSample.wdl) for each indvidual cell and then merging the resulting Zarr output files. 

More details about the Smart-seq2 pipeline can be read in the [Smart-seq2 Single Sample worflow documentation](https://github.com/HumanCellAtlas/skylab/tree/master/pipelines/smartseq2_single_sample) in GitHub. 
 
## Inputs 

The [Smartseq2PlateExample.json](SmartSeq2PlateExample.json) file provides example inputs for the Smart-seq2 Multi Sample workflow. 

### Sample Data Input
The sample data inputs are identifical to those specified in the [Smart-seq2 Single Sample README](https://github.com/HumanCellAtlas/skylab/blob/master/pipelines/smartseq2_single_sample/README.md). The Smart-seq2 Multi Sample workflow processes both single- and paired-end samples; however, these samples can not be mixed in the same run. 
For the Multi Sample workflow, fastq files must be named with the following convention:

*sample_id*\_1.fastq (forward reads for paired-end samples or reads for single-end samples)
*sample_id*\_2.fastq (reverse reads for paired-end samples and not applicable to single-end samples)


### Additional Input
The reference inputs are identical to those specified in the "Additional Reference Inputs" section of the [Smart-seq2 Single Sample README](https://github.com/HumanCellAtlas/skylab/blob/master/pipelines/smartseq2_single_sample/README.md). 

In addition to the Smart-seq2 Single Sample inputs, the multi sample workflow requires the inputs listed in the following table.

| Input name | Input Description | Input Type |
| --- | --- | --- |
| file_prefix | The URL path to the fastq file folder | String | 
| input_file_names | Array of filename prefixes; will be appended with \_1.fastq.gz and \_2.fastq.gz | Array of Strings |
| bacth_id | Identifier for the batch of multiple samples | String |
| output_loom | boolean description for optional loom output | Boolean |


The "input_file_names" is an array containing only the prefix for each fastq file. For example, if you have a paired-end sample called "Sample1", the two fastq file names will be "Sample1_1.fastq" and "Sample1_2.fastq". The input_file_names array would only use the prefix "Sample1" in addition to the other sample name prefixes. 

 
 ## Smart-seq2 Multi Sample Task Summary
 
 ## Outputs
 
 # Versioning
 
