---
title: "Impact of reference bundle on scRNA-Seq alignment"
output:
  html_document: default
  html_notebook: default
---
This the first part of analysis about the impact of reference bundle on scRNA-Seq alignment and gene quantifications. Alignment metrics have been parsed from `STAR` alignment log file `Final.out.log`. The metrics we will examine is `unique mapping reads`,`multiple mapping reads`, `reads mapped too many loci`, `total number of annotated junctions`. 

First,load packages `ggplot2` and `plyr`

```{r}
library(ggplot2)
library(plyr)
```

Now take a look at the `STAR` logs. 

```{r}
ref_log<-read.delim('~/Documents/HCA/reference/ref_run_MGH30_logs.tsv',sep='\t',header=T)
head(ref_log)
```

Then add string `bp` to column `length` and then rename of `ref_name` 

```{r}
rlen<-paste(ref_log$length,'bp')
refname<-mapvalues(ref_log$ref_name,from=c('GRCh38_Ensembl','GRCh38_GencodeV27','GRCh38_GencodeV27_basic','GRCh38_RefSeq'),to=c('Ensembl','GencodeV27_comp','GencodeV27_basic','RefSeq'))
multi.loci<-unlist(lapply(ref_log$X..of.reads.mapped.to.multiple.loci, function(x) as.numeric(sub("%", "", x))))
ref_log$ref_name<-refname
ref_log$length<-rlen
```

Then let's plot `uniquely mapped reads vs reference bundle`.  `Ensemnl` show the highest unique mapping rate among all 5 reference bundle
```{r}
uniq.mapped<-unlist(lapply(ref_log$Uniquely.mapped.reads.., function(x) as.numeric(sub("%", "", x))))
p<-ggplot(data=ref_log)+geom_point(mapping=aes(x=refname,y=uniq.mapped,color=rlen),size=4)
p<-p+xlab('Reference Bundle')+ylab('% Reads uniquely mapped')+ggtitle('uniquely mapping reads vs reference bundle')
p<-p+theme(plot.title = element_text(hjust = 0.5,size=20,face='bold'))
p<-p+theme(axis.title.x = element_text(color='black',size=10,face='bold'),axis.title.y = element_text(size=10,color='black',face='bold'))
p<-p+theme(axis.text.y = element_text(color='black',size=10,face='bold'),axis.text.x = element_text(color='black',size=10,face='bold',hjust=1,angle = 45))
p<-p+theme(legend.text=element_text(size=12,face='bold'),legend.title = element_text(face="bold"))
p
```

Now let's plot `multiple mapping reads vs reference bundle`.  `Ensembl` bundle actually show the lowest multiple mapping rate

```{r}
multi.loci<-unlist(lapply(ref_log$X..of.reads.mapped.to.multiple.loci, function(x) as.numeric(sub("%", "", x))))
p<-ggplot(data=ref_log)+geom_point(mapping=aes(x=ref_name,y=multi.loci,color=length),size=4)
p<-p+xlab('Reference Bundle')+ylab('% Reads mapped to multiple loci')+ggtitle('multiple mapping reads vs reference bundle')
p<-p+theme(plot.title = element_text(hjust = 0.5,size=20,face='bold'))
p<-p+theme(axis.title.x = element_text(color='black',size=10,face='bold'),axis.title.y = element_text(size=10,color='black',face='bold'))
p<-p+theme(axis.text.y = element_text(color='black',size=10,face='bold'),axis.text.x = element_text(color='black',size=10,face='bold',hjust=1,angle = 45))
p<-p+theme(legend.text=element_text(size=12,face='bold'),legend.title = element_text(face="bold"))
p
```

Then `reads mapped too many loci vs reference bundle`. `Ensembl` bundle show slightly low rate of `reads mapped too many loci`

```{r}
too.many<-unlist(lapply(ref_log$X..of.reads.mapped.to.too.many.loci, function(x) as.numeric(sub("%", "", x))))
p<-ggplot(data=ref_log)+geom_point(mapping=aes(x=refname,y=too.many,color=rlen),size=4)
p<-p+xlab('Reference Bundle')+ylab('% Reads mapped too many loci')+ggtitle('Reads mapped too many loci vs reference bundle')
p<-p+theme(plot.title = element_text(hjust = 0.5,size=20,face='bold'))
p<-p+theme(axis.title.x = element_text(color='black',size=10,face='bold'),axis.title.y = element_text(size=10,color='black',face='bold'))
p<-p+theme(axis.text.y = element_text(color='black',size=10,face='bold'),axis.text.x = element_text(color='black',size=10,face='bold',hjust=1,angle = 45))
p<-p+theme(legend.text=element_text(size=12,face='bold'),legend.title = element_text(face="bold"))
p
```

What if the `annotated splicing junction` ? 

```{r}
junction<-ref_log$Number.of.splices..Annotated..sjdb.
p<-ggplot(data=ref_log)+geom_point(mapping=aes(x=refname,y=junction,color=rlen),size=4)
p<-p+xlab('Reference Bundle')+ylab('# of annotated junction')+ggtitle('Annotated junction vs reference bundle')
p<-p+theme(plot.title = element_text(hjust = 0.5,size=20,face='bold'))
p<-p+theme(axis.title.x = element_text(color='black',size=10,face='bold'),axis.title.y = element_text(size=10,color='black',face='bold'))
p<-p+theme(axis.text.y = element_text(color='black',size=10,face='bold'),axis.text.x = element_text(color='black',size=10,face='bold',hjust=1,angle = 45))
p<-p+theme(legend.text=element_text(size=12,face='bold'),legend.title = element_text(face="bold"))
p
```

`total splices` include `annotated` and `not annotated` junctions
```{r}
splices<-ref_log$Number.of.splices..Total
p<-ggplot(data=ref_log)+geom_point(mapping=aes(x=refname,y=junction,color=rlen),size=4)
p<-p+xlab('Reference Bundle')+ylab('# of Total Splices')+ggtitle('Total Splices vs reference bundle')
p<-p+theme(plot.title = element_text(hjust = 0.5,size=20,face='bold'))
p<-p+theme(axis.title.x = element_text(color='black',size=10,face='bold'),axis.title.y = element_text(size=10,color='black',face='bold'))
p<-p+theme(axis.text.y = element_text(color='black',size=10,face='bold'),axis.text.x = element_text(color='black',size=10,face='bold',hjust=1,angle = 45))
p<-p+theme(legend.text=element_text(size=12,face='bold'),legend.title = element_text(face="bold"))
p
```
