| Pipeline Version | Date Updated | Documentation Author | Questions or Feedback |
| :----: | :---: | :----: | :--------------: |
| [smartseq2_v2.5.0](https://github.com/HumanCellAtlas/skylab/releases/tag/smartseq2_v2.5.0) | November 07, 2019 | [Elizabeth Kiernan](mailto:ekiernan@broadinstitute.org) | Please file GitHub issues in skylab or contact [Kylee Degatano](mailto:kdegatano@broadinstitute.org) |

# Table of Contents


# Smart-Seq2 Single Sample Pipeline Overview

![](pipeline_two_tracks.png)

## Introduction to the Smart-seq2 Pipeline

The Smart-seq 2 Single Sample pipeline is designed by the [Data Coordination Platform](https://data.humancellatlas.org/about) of the [Human Cell Atlas](https://www.humancellatlas.org/) to process single-cell RNAseq (scRNAseq) data generated by [Smart-seq2 assays](https://www.nature.com/articles/nmeth.2639). The pipeline's workflow is written in WDL, is freely available on [GitHub](https://github.com/HumanCellAtlas/skylab), and can be run by any compliant WDL runner (e.g. [crowmell](https://github.com/broadinstitute/cromwell)). The pipeline is designed to process stranded or un-stranded, paired- or single-end, scRNA-seq data from an invididual cell (to process multiple cells, see the [smartseq2_multisample](https://github.com/HumanCellAtlas/skylab/tree/master/pipelines/smartseq2_multisample) pipeline). 

The Smartseq 2 Single Sample workflow is divided into two parts. First, it aligns reads to the genome and performs quality control on genome-aligned BAMs. Second, it aligns reads to the transcriptome and generates gene expression estimates from transcriptome-aligned BAMs. The pipeline returns reads in BAM and an optional Zarr format, and it returns QC metrics in CSV format. 

This pipeline is validated to support both human and mouse datasets.   

## Quick Start Table

| Pipeline Features | Description | Source |
|-------------------|---------------------------------------------------------------|-----------------------|
| Assay Type | Smart-seq2 Single Sample | [Smart-seq2](https://www.nature.com/articles/nprot.2014.006)
| Overall Workflow  | Quality control module and transcriptome quantification module | Code available from [Github](https://github.com/HumanCellAtlas/skylab/blob/master/pipelines/smartseq2_single_sample/SmartSeq2SingleSample.wdl) |
| Workflow Language | WDL | [openWDL](https://github.com/openwdl/wdl) |
| Genomic Reference Sequence (for validation)| GRCh38 human genome primary sequence and M21 (GRCm38.p6) mouse genome primary sequence | GENCODE [Human](https://www.gencodegenes.org/human/release_27.html) and [Mouse](https://www.gencodegenes.org/mouse/release_M21.html) 
| Transcriptomic Reference Annotation (for validation) | V27 GENCODE human transcriptome and M21 mouse transcriptome | GENCODE [Human](ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_27/gencode.v27.annotation.gtf.gz) and [Mouse](ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M21/gencode.vM21.annotation.gff3.gz) |
| Aligner  | HISAT2 (v.2.1.0) | [Kim, et al.,2019](https://www.nature.com/articles/s41587-019-0201-4) |
| QC Metrics | Picard (v.2.10.10) | Broad Institute [(https://broadinstitute.github.io/picard/)](https://broadinstitute.github.io/picard/)   |
| Transcript Quantification | Utilities for processing large-scale single cell datasets | [RSEM v.1.3.0](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-12-323)                          
| Data Input File Format | File format in which sequencing data is provided | [FASTQ](https://academic.oup.com/nar/article/38/6/1767/3112533) |                     
| Data Output File Formats | File formats in which Smart-seq2 output is provided | [BAM](http://samtools.github.io/hts-specs/), [Zarr version 2](https://zarr.readthedocs.io/en/stable/spec/v2.html) (optional output), CSV (QC Metrics) |

# Set-Up

## Smart-seq2 Installation and Requirements

The Smart-seq2 workflow code can be downloaded by cloning the GitHub repository skylab. For the latest release of Smart-seq2, please see the [release tags prefixed with "Smartseq2"](https://github.com/HumanCellAtlas/skylab/releases).

The workflow is deployed using [Cromwell](https://github.com/broadinstitute/cromwell), a GA4GH compliant, flexible workflow management system that supports multiple computing platforms. 

## Inputs

Example Smart-seq2 Single Sample inputs are found in the [SmartSeq2SingleSampleExample.json](SmartSeq2SingleSampleExample.json) file. 

### Sample data input

The pipeline is designed for both single- and paired-end reads in the form of fastq files. The workflow accepts two fastq files for paired-end experiments and one fastq file single-end experiments. It processes one sample (cell) at a time.  

*  fastq1: forward reads for sample with paired-end sequencing (or reads for sample with single-end sequencing)
*  fastq2: reverse reads for sample with paired-end sequencing (not applicable for samples with single-end sequencing)

### Additional Reference Inputs:
The Smartseq2 Single Sample workflow requires multiple reference indexes. Information on how these indexes are created is found in the [Creating_Smartseq2_References]() documentation. The table below describes the references and each tool that uses the reference as input.

| Workflow Step | Reference name | Reference Description | Tool |
| --- | --- | --- | --- |
| All | sample_name | Sample name or Cell ID | NA |
| All | output_name | Output name, can include path | NA |
| **Genomic alignment with HISAT2** | hisat2_ref_index | HISAT2 reference index file in tarball | HISAT2 |
|     | hisat2_ref_name | HISAT2 reference index name | HISAT2 |
| **Picard-generated Quality Control Metrics** | genome_ref_fasta | Genome reference in fasta format | Picard |
|     | gene_ref_flat | [RefFlat](https://software.broadinstitute.org/software/igv/genePred) file containing the location of RNA transcripts, exon start sites, etc.  | Picard |
|     | rrna_intervals | Ribosomal RNA intervals file | Picard |
|     | stranded | Library strand information for HISAT2; example values include FR (read corresponds to transcript), RF(read corresponds to reverse compliment of transcript), or NONE | Picard |
| **Transcriptomic Alignment with HISAT2** | hisat2_ref_trans_index | HISAT2 transcriptome index file in tarball | HISAT2 |
|     | hisat2_ref_trans_name | HISAT2 transcriptome index file name | HISAT2 |
| **Gene Expression Quantificaiton with RSEM** | rsem_ref_index | RSEM reference index file in tarball |

# Running Smart-seq2

The [SmartSeq2SingleSample.wdl](SmartSeq2SingleSample.wdl) is in the [pipelines/smartseq2_single_sample folder](https://github.com/HumanCellAtlas/skylab/tree/master/pipelines/smartseq2_single_sample) of the HCA skylab repository and implements the workflow by importing individual tasks (written in WDL script) from the skylab [Library](https://github.com/HumanCellAtlas/skylab/tree/master/library).

## Smarts-seq2 Task Summary

Here we describe the tasks of the Smartseq2 Single Sample pipeline; [the code](SmartSeq2SingleSample.wdl) and [library of tasks](https://github.com/HumanCellAtlas/skylab/tree/master/library) are available through GitHub.

Overall, the workflow has two parts:

**Part 1: Quality Control Tasks**
 1. Aligns reads to the genome with HISAT2 v.2.1.0
 2. Calculates summary metrics from an aligned BAM using Picard

**Part 2: Transcriptome Quantification Tasks**
 1. Aligns reads to the transcriptome with HISAT v.2.1.0
 2. Quantifies gene expression using RSEM v.1.3.0

### Part 1: Quality Control Tasks
#### 1.1 Align reads using HISAT2
The Smartseq2 Single Sample workflow uses the [HISAT2.wdl task](https://github.com/HumanCellAtlas/skylab/blob/master/library/tasks/HISAT2.wdl) to perform graph-based alignment of paired- or single-end reads (in the form of fastq files) to a reference genome. This task uses HISAT2, a fast-paced, cost-efficient alignment tool. For reference input, HISAT2 requires an index which can be built following the [HCA's build_indices documentation](https://github.com/HumanCellAtlas/skylab/tree/master/library/accessory_workflows/build_indices). HISAT2 can determine the presence of non-transcript sequences and true transcript sequences, taking into account the presence of single-nucleotide polymorphisms by using references such as NCBI's [dbSNP Build 150](https://www.ncbi.nlm.nih.gov/projects/SNP/snp_summary.cgi?view+summary=view+summary&build_id=150). The HISAT2 outputs are an aligned bam file, a bam index, and an alignment log file. 

#### 1.2 Calculate summary metrics using Picard
Quality control measurements are calculated with [Picard tools](https://broadinstitute.github.io/picard/), command line tools used for manipulating high-throughput sequencing data. Picard uses the aligned BAM file generated with HISAT2 as input. Additionally, it requires a reference genome fasta, a gene refflat, and an RNA intervals file (see the [Creating_Smartseq2_References]() documentation). 

The [Picard.wdl](https://github.com/HumanCellAtlas/skylab/blob/master/library/tasks/Picard.wdl) generates QC metrics by using three tasks:

*  CollectMultipleMetrics: calls the  [CollectMultipleMetrics](https://software.broadinstitute.org/gatk/documentation/tooldocs/4.0.0.0/picard_analysis_CollectMultipleMetrics.php) tool which uses the aligned BAM file and reference genome fasta to collect metrics on [alignment](http://broadinstitute.github.io/picard/picard-metric-definitions.html#AlignmentSummaryMetrics), [insert size](http://broadinstitute.github.io/picard/picard-metric-definitions.html#InsertSizeMetrics), [GC bias](https://broadinstitute.github.io/picard/command-line-overview.html#CollectGcBiasMetrics), [base distribution by cycle](http://broadinstitute.github.io/picard/picard-metric-definitions.html#BaseDistributionByCycleMetrics), [quality score distribution](https://broadinstitute.github.io/picard/command-line-overview.html#QualityScoreDistribution), [quality distribution by cycle](https://broadinstitute.github.io/picard/command-line-overview.html#MeanQualityByCycle), [sequencing artifacts](http://broadinstitute.github.io/picard/picard-metric-definitions.html#ErrorSummaryMetrics), and [quality yield](http://broadinstitute.github.io/picard/picard-metric-definitions.html#CollectQualityYieldMetrics.QualityYieldMetrics). 

*  CollectRnaMetrics: calls the [CollectRnaSeqMetrics](https://software.broadinstitute.org/gatk/documentation/tooldocs/4.0.0.0/picard_analysis_CollectRnaSeqMetrics.php) tool which uses the aligned BAM, a RefFlat genome annotation file, and a ribosomal intervals file to produce RNA alignment metrics (metric descriptions are found in the [Picard Metrics Dictionary](http://broadinstitute.github.io/picard/picard-metric-definitions.html#RnaSeqMetrics). 

*  CollectDuplicationMetrics: calls the [MarkDuplicates]((https://software.broadinstitute.org/gatk/documentation/tooldocs/4.0.4.0/picard_sam_markduplicates_MarkDuplicates.php) tool which uses the aligned BAM to identify duplicate reads (output metrics are listed in the [Picard Metrics Dictionary](http://broadinstitute.github.io/picard/picard-metric-definitions.html#DuplicationMetrics).


### Part 2: SmartSeq2 scRNASeq Quantification
The pipeline uses RSEM to generate expression estimates from the SAM/BAM files generated by HISAT2 alignment of reads against a transcriptome. The **quantification** steps have been incorperated into the SmartSeq2 pipeline.

#### 1. Align reads to the transcriptome


#### 2. Quantify gene expression using RSEM
The quantification of genes and transcripts is done with the [RSEM](https://deweylab.github.io/RSEM/rsem-calculate-expression.html) program. The program `rsem-calculate-expression` in the RSEM package is used to estimate gene/isoform expression levels. The output can be `expected_counts`, `TPM` or `FPKM`. 

#### Create the RSEM Index
RSEM provides a program to extract and pre-process the reference sequences, then build its index. The workflow to build the RSEM index is available at [Github](https://github.com/HumanCellAtlas/skylab/blob/master/pipelines/accessory_workflows/build_rsem_reference/BuildRsemReferenceBundle.wdl). To learn more about how to build the RSEM index, refer to the [RSEM website](http://deweylab.biostat.wisc.edu/rsem/rsem-prepare-reference.html#description).
**Note** A transcriptome reference Fasta file is included in RSEM index bundle. This file will be used to create the HISAT2 transcriptome index.

# Outputs

# Versioning
# Pipeline Improvements
